public/components/explorer/visualizations/drag_drop/drag_drop.tsx

```diff
@@ -0,0 +1,255 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+import './drag_drop.scss';
+
+import React, { useState, useContext } from 'react';
+import classNames from 'classnames';
+import { DragContext } from './providers';
+// import { trackUiEvent } from '../lens_ui_telemetry';
+
+type DroppableEvent = React.DragEvent<HTMLElement>;
+
+/**
+ * A function that handles a drop event.
+ */
+export type DropHandler = (item: unknown) => void;
+
+/**
+ * The base props to the DragDrop component.
+ */
+interface BaseProps {
+  /**
+   * The CSS class(es) for the root element.
+   */
+  className?: string;
+
+  /**
+   * The event handler that fires when an item
+   * is dropped onto this DragDrop component.
+   */
+  onDrop?: DropHandler;
+
+  /**
+   * The value associated with this item, if it is draggable.
+   * If this component is dragged, this will be the value of
+   * "dragging" in the root drag/drop context.
+   */
+  value?: unknown;
+
+  /**
+   * Optional comparison function to check whether a value is the dragged one
+   */
+  isValueEqual?: (value1: unknown, value2: unknown) => boolean;
+
+  /**
+   * The React element which will be passed the draggable handlers
+   */
+  children: React.ReactElement;
+
+  /**
+   * Indicates whether or not the currently dragged item
+   * can be dropped onto this component.
+   */
+  droppable?: boolean;
+
+  /**
+   * Additional class names to apply when another element is over the drop target
+   */
+  getAdditionalClassesOnEnter?: () => string;
+
+  /**
+   * The optional test subject associated with this DOM element.
+   */
+  'data-test-subj'?: string;
+
+  /**
+   * Indicates to the user whether the currently dragged item
+   * will be moved or copied
+   */
+  dragType?: 'copy' | 'move';
+
+  /**
+   * Indicates to the user whether the drop action will
+   * replace something that is existing or add a new one
+   */
+  dropType?: 'add' | 'replace';
+}
+
+/**
+ * The props for a draggable instance of that component.
+ */
+interface DraggableProps extends BaseProps {
+  /**
+   * Indicates whether or not this component is draggable.
+   */
+  draggable: true;
+  /**
+   * The label, which should be attached to the drag event, and which will e.g.
+   * be used if the element will be dropped into a text field.
+   */
+  label: string;
+}
+
+/**
+ * The props for a non-draggable instance of that component.
+ */
+interface NonDraggableProps extends BaseProps {
+  /**
+   * Indicates whether or not this component is draggable.
+   */
+  draggable?: false;
+}
+
+type Props = DraggableProps | NonDraggableProps;
+
+/**
+ * A draggable / droppable item. Items can be both draggable and droppable at
+ * the same time.
+ *
+ * @param props
+ */
+
+export const DragDrop = (props: Props) => {
```

Comment:
i think there are existing react libraries that can handle drag drop, any reason to implement it instead?

Reply:
I was borrowing code directly from kibana lens, since the design of visualization in explorer looks similar to it and sharing some of the functionalities. But moving forward, I'll take look at react drag and drop libraries to see what could be the best solution for this feature of visualization.